File: backend/aspnetcore_backend/Core.Domain/ValueObjects/Cielab.cs
using System;

namespace Core.Domain.ValueObjects
{
    public class Cielab
    {
        public float Ligthness { get; }
        public float aAxis { get; }
        public float bAxis { get; }

        public Cielab(float lightness, float aAxis, float bAxisValue)
        {
            this.Ligthness = lightness;
            this.aAxis = aAxis;
            this.bAxis = bAxis;
        }

        public RGB ToRGB() => throw new NotImplementedException();
        public Hexadecimal ToHexadecimal() => this.ToRGB().ToHex();

        internal double CalculateCie94Distance(Cielab cielabColor)
        {
            double deltaL = this.Ligthness - cielabColor.Ligthness;

            double C1 = Math.Sqrt(Math.Pow(this.aAxis, 2) + Math.Pow(cielabColor.aAxis, 2));
            double C2 = Math.Sqrt(Math.Pow(this.bAxis, 2) + Math.Pow(cielabColor.bAxis, 2));
            double deltaC = C1 - C2;

            double deltaA = this.aAxis - cielabColor.aAxis;
            double deltaB = this.bAxis - cielabColor.bAxis;
            double deltaH = Math.Sqrt(Math.Pow(deltaA, 2) + Math.Pow(deltaB, 2) + Math.Pow(deltaC, 2));

            double kSubL = 1;
            double kSubC = 1;
            double kSubH = 1;
            double K1 = 0.045;
            double K2 = 0.015;
            double sSubL = 1;
            double sSubC = 1 + K1*C1;
            double sSubH = 1 + K2*C1;

            double deltaLQuotient = deltaL/(kSubL*sSubL);
            double deltaCQuotient = deltaL/(kSubC*sSubC);
            double deltaHQuotient = deltaL/(kSubH*sSubH);
            
            double deltaE = Math.Sqrt(Math.Pow(deltaLQuotient, 2) + Math.Pow(deltaCQuotient, 2) + Math.Pow(deltaHQuotient, 2));
            return deltaE;
        }
    }
}

File: backend/source/paints/colorSystem/Cielab.ts
import { Hexadecimal } from "./Hexadecimal";
import { RGB } from "./RGB";
import { XYZComponents } from "./XYZComponents";

export class Cielab {
    public constructor(
        readonly Lightness: number, 
        readonly AAxisValue: number, 
        readonly BAxisValue: number
    ) { }

    public ToRGB(): RGB {
        const xyzComponents = XYZComponents.FromCielab(this);
        const [redComponent, greenComponent, blueComponent] = this.CalculateRGBComponentsFromXYZComponents(xyzComponents);

        return new RGB(redComponent, greenComponent, blueComponent)
    }

    public ToHex(): Hexadecimal {
        return this.ToRGB().ToHex();
    }

    private CalculateRGBComponentsFromXYZComponents({X, Y, Z}: XYZComponents): [number, number, number] {
        let redComponent = X *  3.2406 + Y * -1.5372 + Z * -0.4986;
        let greenComponent = X * -0.9689 + Y *  1.8758 + Y *  0.0415;
        let blueComponent = X *  0.0557 + Y * -0.2040 + Z *  1.0570;
        redComponent = (redComponent > 0.0031308) ? (1.055 * Math.pow(redComponent, 1/2.4) - 0.055) : 12.92 * redComponent;
        greenComponent = (greenComponent > 0.0031308) ? (1.055 * Math.pow(greenComponent, 1/2.4) - 0.055) : 12.92 * greenComponent;
        blueComponent = (blueComponent > 0.0031308) ? (1.055 * Math.pow(blueComponent, 1/2.4) - 0.055) : 12.92 * blueComponent;

        return [redComponent, greenComponent, blueComponent];
    }

    public CalculateCie94Distance(color: Cielab): number {
        const deltaL = this.Lightness - color.Lightness;
        const C1 = Math.sqrt(Math.pow(this.AAxisValue, 2) + Math.pow(color.AAxisValue, 2));
        const C2 = Math.sqrt(Math.pow(this.BAxisValue, 2) + Math.pow(color.BAxisValue, 2));
        const deltaC = C1 - C2;
        const deltaA = this.AAxisValue - color.AAxisValue;
        const deltaB = this.BAxisValue - color.BAxisValue;
        const deltaH = Math.sqrt(Math.pow(deltaA, 2) + Math.pow(deltaB, 2) + Math.pow(deltaC, 2));
        
        const kSubL = 1;
        const kSubC = 1;
        const kSubH = 1;
        const K1 = 0.045;
        const K2 = 0.015;
        const sSubL = 1;
        const sSubC = 1 + K1*C1;
        const sSubH = 1 + K2*C1;
        
        const deltaEDeltaLQuotient = deltaL/(kSubL*sSubL);
        const deltaEDeltaCQuotient = deltaC/(kSubC*sSubC);
        const deltaEDeltaHQuotient = deltaH/(kSubH*sSubH);
        const deltaE = Math.sqrt(Math.pow(deltaEDeltaLQuotient, 2) + Math.pow(deltaEDeltaCQuotient, 2) + Math.pow(deltaEDeltaHQuotient, 2));
        return deltaE;
    }
}